### PR TITLE
Set concurrency limit

### PR DESCRIPTION
**Pull request by Wwwsylvia**
Refactor the worker package to limit amount of concurrent DELETE requests to ACR server

### CODE REVIEW

File: cmd/worker/purger.go [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+	"github.com/Azure/acr-cli/cmd/api"
+)
+
+// Purger defines a worker pool that manages workers to process PurgeJobs.
+type Purger struct {
+	wp *workerPool
+}
+
+// NewPurger creates a new Purger.
+func NewPurger(batchSize int, acrClient api.AcrCLIClientInterface) *Purger {
```
Comment (shizhMSFT) [inline cmd/worker/purger.go:17]: >NewPurger [](start = 5, length = 9)

Users should be able to control the size of the worker pool.

File: cmd/worker/purger.go [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+	procNum := int64(runtime.GOMAXPROCS(0))
+	worker := newPurgeWorker(acrClient)
```
Comment (shizhMSFT) [inline cmd/worker/purger.go:19]: >newPurgeWorker [](start = 11, length = 14)

It is not a worker. It is a job

**Comment by shizhMSFT** [comment]
> type PurgeJob struct {

Can you divide it into two structs as `PurgeTagJob` and `PurgeManifestJob`, having a common `PurgeJobBase` and matches a `PurgeJob` interface?

---
Refers to: cmd/worker/purge_structures.go:9 in 3a71145. [](commit_id = 3a71145d24ea98cee255b97b299c208b72060f94, deletion_comment = False)

File: cmd/acr/purge.go [hunk 3]
```diff
@@ -158,7 +151,7 @@ func newPurgeCmd(out io.Writer, rootParams *rootParameters) *cobra.Command {
 }
 
 // purgeTags deletes all tags that are older than the ago value and that match the tagFilter string.
-func purgeTags(ctx context.Context, acrClient api.AcrCLIClientInterface, loginURL string, repoName string, ago string, tagFilter string, keep int) (int, error) {
+func purgeTags(ctx context.Context, acrClient api.AcrCLIClientInterface, p *worker.Purger, loginURL string, repoName string, ago string, tagFilter string, keep int) (int, error) {
```
Comment (shizhMSFT) [inline cmd/acr/purge.go:154]: >p [](start = 73, length = 1)

have a better variable name

File: cmd/acr/purge.go [hunk 4]
```diff
@@ -185,18 +178,16 @@ func purgeTags(ctx context.Context, acrClient api.AcrCLIClientInterface, loginUR
...
 			}
 			// To not overflow the error channel capacity the purgeTags function waits for a whole block of
-			// 100 jobs to be finished before continuing.
-			wg.Wait()
-			for len(worker.ErrorChannel) > 0 {
-				wErr := <-worker.ErrorChannel
-				if wErr.Error != nil {
-					return -1, wErr.Error
+			// jobs to be finished before continuing.
+			p.Wait(ctx)
```
Comment (shizhMSFT) [inline cmd/acr/purge.go:186]: There will be a dead lock if there are more tags than the batch size.

File: cmd/worker/workerpool.go [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+	}
+
+	go func() {
+		defer wp.sem.Release(1)
+		wp.errChan <- wp.worker.work(ctx, job)
+	}()
+}
+
+// wait waits for all the workers to finish.
+func (wp *workerPool) wait(ctx context.Context) {
```
Comment (shizhMSFT) [inline cmd/worker/workerpool.go:48]: You should use a waitgroup to wait.

File: cmd/worker/purgejob.go [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+
+// process calls acrClient to delete a manifest
+func (job *purgeManifestJob) process(ctx context.Context, acrClient api.AcrCLIClientInterface) error {
+	resp, err := acrClient.DeleteManifest(ctx, job.repoName, job.digest)
+	if err == nil {
+		fmt.Printf("Deleted %s/%s@%s\n", job.loginURL, job.repoName, job.digest)
+		return nil
+	}
+
+	if resp != nil && resp.Response != nil && resp.StatusCode == http.StatusNotFound {
```
Comment (shizhMSFT) [inline cmd/worker/purgejob.go:69]: >resp.StatusCode [](start = 43, length = 15)

need special nil check

File: cmd/worker/purgejob.go [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+
+// process calls acrClient to delete a tag
+func (job *purgeTagJob) process(ctx context.Context, acrClient api.AcrCLIClientInterface) error {
+	resp, err := acrClient.DeleteAcrTag(ctx, job.repoName, job.tag)
+	if err == nil {
+		fmt.Printf("Deleted %s/%s:%s\n", job.loginURL, job.repoName, job.tag)
+		return nil
+	}
+
+	if resp != nil && resp.Response != nil && resp.StatusCode == http.StatusNotFound {
```
Comment (shizhMSFT) [inline cmd/worker/purgejob.go:86]: >resp.StatusCode [](start = 43, length = 15)

same here

File: cmd/worker/purger.go [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+}
+
+// NewPurger creates a new Purger.
+func NewPurger(workerNum int, batchSize int, acrClient api.AcrCLIClientInterface) *Purger {
+	workerPool := newWorkerPool(workerNum, batchSize)
+	return &Purger{
+		workerPool: workerPool,
+		acrClient:  acrClient,
+		wg:         sync.WaitGroup{},
+	}
```
Comment (shizhMSFT) [inline cmd/worker/purger.go:24]: nit: no need to initialize

File: cmd/worker/purger.go [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+// BatchSize returns the size of a batch of PurgeJobs.
+func (p *Purger) BatchSize() int {
+	return p.workerPool.batchSize
+}
+
+// StartPurgeManifest starts a purge manifest job in worker pool.
+func (p *Purger) StartPurgeManifest(ctx context.Context, loginURL string, repoName string, digest string) {
+	job := newPurgeManifestJob(loginURL, repoName, digest)
+
+	p.workerPool.start(ctx, job, p.acrClient, &p.wg)
```
Comment (shizhMSFT) [inline cmd/worker/purger.go:41]: >&p.wg [](start = 43, length = 5)

Please keep the wait group inside purger and not pass to the worker pool

File: cmd/acr/purge.go [hunk 10]
```diff
@@ -151,14 +163,15 @@ func newPurgeCmd(out io.Writer, rootParams *rootParameters) *cobra.Command {
 	cmd.Flags().IntVar(&purgeParams.keep, "keep", 0, "Number of latest to-be-deleted tags to keep, use this when you want to keep at least x number of latest tags that could be deleted meeting all other filter criteria")
 	cmd.Flags().StringArrayVarP(&purgeParams.filters, "filter", "f", nil, "Specify the repository and a regular expression filter for the tag name, if a tag matches the filter and is older than the duration specified in ago it will be deleted")
 	cmd.Flags().StringArrayVarP(&purgeParams.configs, "config", "c", nil, "Authentication config paths (e.g. C://Users/docker/config.json)")
+	cmd.Flags().IntVarP(&purgeParams.taskNumber, "task-number", "t", 0, taskNumberDescription)
```
Comment (shizhMSFT) [inline cmd/acr/purge.go:166]: >0 [](start = 66, length = 1)

default value should be set here.

File: cmd/acr/purge.go [hunk 11]
```diff
@@ -184,20 +197,24 @@ func purgeTags(ctx context.Context, acrClient api.AcrCLIClientInterface, loginUR
...
-				wErr := <-worker.ErrorChannel
-				if wErr.Error != nil {
-					return -1, wErr.Error
-				}
+			// jobs to be finished before continuing.
+			purger.Wait()
+			err = purger.FlushErrChan()
+			if err != nil {
+				return -1, err
 			}
```
Comment (shizhMSFT) [inline cmd/acr/purge.go:218]: We probably need to re-design this part. The objective is that we should return error if there is an error.

**Review by shizhMSFT — CHANGES_REQUESTED** [review]
:clock1:

File: cmd/acr/purge.go [hunk 12]
```diff
@@ -42,27 +41,31 @@ const (
   - Delete all tags older than 1 day in the example.azurecr.io registry inside the hello-world repository using the credentials found in 
     the C://Users/docker/config.json path
 	acr purge -r example --filter "hello-world:.*" --ago 1d --config C://Users/docker/config.json
-`
 
-	defaultNumWorkers       = 6
+  - Delete all tags older than 1 day in the example.azurecr.io registry inside the hello-world repository, with 4 purge tasks running concurrently
+	acr purge -r example --filter "hello-world:.*" --ago 1d --task-number 4
```
Comment (shizhMSFT) [inline cmd/acr/purge.go:46]: >--task-number [](start = 57, length = 13)

The name `task-number` is so weird. Here are some suggestions:
- `ab` has `-c concurrency`.
- `parallel` has `-j maxjobs`
- `xargs` has `-P max-procs, --max-procs=max-procs`

File: cmd/worker/pool.go [hunk 13]
```diff
@@ -0,0 +1,85 @@
...
+	"sync"
+	"sync/atomic"
+
+	"github.com/Azure/acr-cli/cmd/api"
+	"golang.org/x/sync/semaphore"
+)
+
+// Pool uses semaphore to limit the number of goroutines processing PurgeJob.
+type Pool struct {
+	ctx     context.Context
```
Comment (shizhMSFT) [inline cmd/worker/pool.go:15]: >ctx     context.Context [](start = 1, length = 23)

The `ctx` is assigned but never used.

**Comment by Wwwsylvia** [comment]
@shizhMSFT Please help to review, Thanks!

File: cmd/worker/pool.go [hunk 14]
```diff
@@ -0,0 +1,50 @@
...
+	"context"
+	"sync"
+	"sync/atomic"
+
+	"github.com/Azure/acr-cli/cmd/api"
+)
+
+// pool manages a limited number of workers that process purgeJob.
+type pool struct {
+	size int
```
Comment (shizhMSFT) [inline cmd/worker/pool.go:13]: >size int [](start = 1, length = 8)

It seems redundant.

**Review by shizhMSFT — APPROVED** [review]
:shipit:

**Pull request merged**